### ISSUE TITLE
Each CullerRenderer creates a new WebGL context

### ISSUE THREAD

**Issue by onderilkesever** [ThatOpen/engine_components#464]
### Describe the bug 📝

Each CullerRenderer uses one WebGL context. Browsers have hard limits to the number of contexts available (typically 16). This means that an application may only have 16 Cullers at the same time. This becomes a limiting factor in cases we have more than one viewport in a scene.

When the WebGL context limit has been reached, the browser will remove older WebGL contexts to keep within the hard limit. This will lead to the application using CullerRenderer breaking down or misbehaving.

### System Info 💻

```shell
System:
    OS: macOS 14.4.1
    CPU: (10) arm64 Apple M1 Pro
    Memory: 92.77 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.14.0 - /opt/homebrew/opt/node@18/bin/node
    npm: 9.3.1 - /opt/homebrew/opt/node@18/bin/npm
  Browsers:
    Chrome: 127.0.6533.90
    Edge: 127.0.2651.86
    Safari: 17.4.1
```

### Used Package Manager 📦

npm

### Validations ✅

- [X] Read the [docs](https://docs.thatopen.com/intro).
- [X] Check that there isn't [already an issue](https://github.com/ThatOpen/engine_components/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a repository issue and not a framework-specific issue. For example, if it's a THREE.js related bug, it should likely be reported to [mrdoob/threejs](https://github.com/mrdoob/three.js) instead.
- [X] Check that this is a concrete bug. For Q&A join our [Community](https://people.thatopen.com/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by agviegas** [comment]
@onderilkesever what solution would you suggest?

**Comment by onderilkesever** [comment]
Couple suggestions;

**Comment by onderilkesever** [comment]
@agviegas thanks for quick response. Maybe single renderer can be used for all cullers, but you need to reset it all the time before rendering off screen so i am not sure about the performance.

**Comment by agviegas** [comment]
@onderilkesever we could do that, but the problem is that sharing the same renderer among many cullers might result in the cullers queueing to use it, damaging the experience.

If three.js can't support more than simultaneous 16 renderers, I'd stick with that limit for our library as well for now. Regarding disposing:

> it will fail the next time when trying to create new one.

I just tried this, and it works perfectly. Did you try this on your end? It seems like you can create and dispose as many cullers as you want:

```ts
const cullers = components.get(OBC.Cullers);
let culler = cullers.create(world);
for (let i = 0; i < 100; i++) {
  culler.dispose();
  culler = cullers.create(world);
}
```

**Comment by onderilkesever** [comment]
> If three.js can't support more than 16 simultaneous renderers, I'd stick with that limit for our library as well for now.
It is not the three js what is limiting that number, it's the browsers.. 

> I just tried this, and it works perfectly. Did you try this on your end? It seems like you can create and dispose as many cullers as you want:
I have tried and unfortunately it did not work for me. might be related to that you simply create and dispose, can you maybe simulate real life example e.g add meshes to cullers, update etc. because i think whats eventually not releasing the webgl contexts is that some resources have still been referenced from somewhere, and that leads my garbage collection to not collect those context.

**Comment by agviegas** [comment]
@onderilkesever I was able to reproduce the issue with this code:

```ts
const cullers = components.get(OBC.Cullers);
let culler = cullers.create(world);

let counter = 0;
async function disposeCullerRecursive() {
  if (culler) {
    cullers.delete(world);
  }

  culler = cullers.create(world);
  culler.enabled = true;

  for (const fragment of model.items) {
    culler.add(fragment.mesh);
  }

  culler.onViewUpdated.add(() => {
    if (counter < 100) {
      console.log(counter++);
      disposeCullerRecursive();
    }
  });

  culler.updateVisibility(true);
}

await disposeCullerRecursive();
```

https://github.com/user-attachments/assets/c22def93-8147-464b-a2e3-f38e1dfbbf2d

I discovered that Three.js has [a way to force the release of the context](https://discourse.threejs.org/t/i-have-a-problem-in-warning-too-many-active-webgl-contexts-oldest-context-will-be-lost/41300/2). Applying it to the culler renderer solves the problem:


https://github.com/user-attachments/assets/df297108-8b5c-4ab8-ba38-cc1f99bca955

I've addeed this everywhere in our library where we dispose a renderer, so this should solve the issue for disposing. 

For having more than 16 active renderers, I don't think we can do anything for now, as any solution would reduce the performance and/or flexibility of the culler. If this becomes a major issue for the community, we'll see if we can find a solution.

This fix is available from `@thatopen/components@2.1.24` and `@thatopen/components-front@2.1.26`. Let me know if you still face the issue.
